### PR TITLE
[WIP] Flat Map Writer Step 1 : Numeric Keyed FlatMap writers (generating streams for all keys).

### PR DESCRIPTION
**Pull request by akhilum**
Test plan - (Please fill in how you tested your changes)

1. I have added a test called TestMapColumnWriter that creates a MapBlock and calls writer on the block.
2. In Columnwriters I have routed MAP and FLATMAP to NumbericKeyedFLatMapWriter for development purposes.
3. presto/orc/writer/ColumnWriterFactory.java - This is a utility factory I have added to create column Writers for each key in the Map
4. Please ignore presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapColumnWriter.java this file is unused. I did a git rm but its still in the commit. I will remove it as I go along
5. presto/orc/writer/KeyNode.java is the ColumnWriter implementation for each key in the Map
6. orc/writer/NumericKeyedFlatMapColumnWriter.java is the mail FlatMap implementation that I'm currently working on
7.  Ignore StringKeyedFlatMapColumnWriter.java its not yet implementaed

Currently debugging my running the test TestMapColumnWriter::testMapwithIntegerKey()

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnWriters.java [hunk 1]
```diff
@@ -139,6 +143,39 @@ public static ColumnWriter createColumnWriter(
...
+                        .setCompressionParameters(compressionParameters)
+                        .setOrcEncoding(orcEncoding)
+                        .setHiveStorageTimeZone(hiveStorageTimeZone)
+                        .setStringStatisticsLimit(stringStatisticsLimit)
+                        .setDwrfEncryptors(dwrfEncryptors)
+                        .setMetadataWriter(metadataWriter)
+                        .setIntegerDictionaryEncodingEnabled(integerDictionaryEncodingEnabled);
+                switch (keyType.getOrcTypeKind()) {
+                    case CHAR:
+                        checkArgument(orcEncoding != DWRF, "DWRF does not support %s type", type);
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnWriters.java:164]: FlatMap encoding is only supported by DWRF format. ORC format, does not support flat map encoding.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/KeyNode.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    public KeyNode(int columnIndex, Type type, ColumnWriter valueWriter, CompressionParameters compressionParameters, Optional<DwrfDataEncryptor> dwrfEncryptor, MetadataWriter metadataWriter) {
+        index = columnIndex;
+        this.valueWriter = valueWriter;
+        inMapStream = new BooleanOutputStream(compressionParameters, dwrfEncryptor);
+        this.buffer = new OrcOutputBuffer(compressionParameters, dwrfEncryptor);
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, compressionParameters, dwrfEncryptor);
+    }
+
+    public void add(Block value) {
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/KeyNode.java:56]: this semantic is confusing to me. Block holds a list of value, but this is more of use like a single value. If single value is supposed to be added, we generally use the semantics add(Block value, int position)

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/KeyNode.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        this.inMapStream.writeBoolean(true);
+        valueWriter.writeBlock(value);
+    }
+
+    public void beginRowGroup() {
+        valueWriter.beginRowGroup();
+    }
+
+    @Override
+    public void writeBlock(Block block)
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/KeyNode.java:67]: What is the difference between add vs writeBlock

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/NumericKeyedFlatMapColumnWriter.java [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    private final Optional<DwrfDataEncryptor> dwrfEncryptor;
+    private final ColumnEncoding columnEncoding;
+    private final LongOutputStream lengthStream;
+    private final PresentOutputStream presentStream;
+    private final CompressedMetadataWriter metadataWriter;
+    private final MetadataWriter inputMetadataWriter;
+    private final ColumnWriterFactory valueWriterFactory;
+    private final Long2ObjectMap<KeyNode> keyToKeyNode = new Long2ObjectOpenHashMap<>();
+    private final OrcType orcKeyType;
+    private List<ColumnWriter> valueWriters;
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/NumericKeyedFlatMapColumnWriter.java:70]: This definition is confusing. What is the hierarchy of the classes.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/NumericKeyedFlatMapColumnWriter.java [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapColumnWriter.class).instanceSize();
+    private final int column;
+    private final MapType type;
+    private final Type keyType;
+    private final Type valueType;
+    private final boolean compressed;
+    private final CompressionParameters compressionParameters;
+    private final Optional<DwrfDataEncryptor> dwrfEncryptor;
+    private final ColumnEncoding columnEncoding;
+    private final LongOutputStream lengthStream;
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/NumericKeyedFlatMapColumnWriter.java:63]: Why another length stream inside the NumericKeyedFlatMapColumnWriter

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/NumericKeyedFlatMapColumnWriter.java [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                        break;
+                    case LONG:
+                        long longKey = (long) this.keyType.getObjectValue(PROPERTIES, keysBlock, offset + entry);
+                        keyAsLong = Long.valueOf(longKey);
+                        break;
+                    default:
+                        throw new IllegalStateException("Unexpected value: " + orcKeyType.getOrcTypeKind());
+                }
+
+                BlockBuilder blockBuilder = this.valueType.createBlockBuilder(null, BATCH_ROWS);
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/NumericKeyedFlatMapColumnWriter.java:190]: Why is a FlatMap writer creating block builder ?

File: presto-common/src/main/java/com/facebook/presto/common/type/StandardTypes.java [hunk 7]
```diff
@@ -40,6 +40,7 @@
     public static final String ROW = "row";
     public static final String ARRAY = "array";
     public static final String MAP = "map";
+    public static final String FLATMAP = "flatmap";
```
[the remark this answers is not in the document]
Comment (akhilum) [reply presto-common/src/main/java/com/facebook/presto/common/type/StandardTypes.java:43]: Yes, I will remove the same.

**Review by arunthirupathi — COMMENTED** [review]
I am half way through the review, it will take more time tomorrow to go through these changes.

**Pull request closed without merging**